# Імпорт Pandas

In [103]:
import pandas as pd

# 1. Завантаження та попередній аналіз даних

**Зчитування файлу та створення `DataFrame`**

Файл `csv` з даними буде взято за посиланням з гітхабу.
Попередньо, в гітхабі треба обрати сирий (raw) формат файлу:

In [104]:
url = 'https://raw.githubusercontent.com/OlhaOsypova/Machine-Learning-and-Data-Analytics/main/Topic_2_Pandas/Task/student_data.csv'

Тепер з отриманого файлу можна створити датафрейм, використовуючи метод `read_csv`:

In [105]:
df = pd.read_csv(url) # створення фрейму
df                    # виведення фрейму на екран

Student_identifier_code_for_participation_in_the_survey  \
0                                               ID_2126        
1                                               ID_1860        
2                                               ID_2130        
3                                               ID_2095        
4                                               ID_2044        
...                                                 ...        
1195                                            ID_1826        
1196                                            ID_1726        
1197                                                NaN        
1198                                            ID_1989        
1199                                            ID_1371        

      Age_of_student_at_the_time_of_the_data_collection_process  \
0                                                  24.0           
1                                                  20.0           
2                                                  19.0           
3                                                  20.0           
4                                                  19.0           
...                                                 ...           
1195                                               22.0           
1196                                               24.0           
1197                                               23.0           
1198                                               21.0           
1199                                               19.0           

     Grade_level_of_the_participant_in_the_educational_program  \
0                                              Freshman          
1                                                Senior          
2                                                Junior          
3                                             Sophomore          
4                                                Senior          
...                                                 ...          
1195                                          Sophomore          
1196                                          Sophomore          
1197                                             Junior          
1198                                          Sophomore          
1199                                          Sophomore          

      Hours_spent_studying_on_a_weekly_basis  \
0                                       13.0   
1                                       28.0   
2                                       21.0   
3                                       28.0   
4                                       25.0   
...                                      ...   
1195                                    36.0   
1196                                    19.0   
1197                                    39.0   
1198                                     7.0   
1199                                    20.0   

     Preferred_study_method_in_lecture_or_practical_sessions  \
0                                               Lecture        
1                                               Lecture        
2                                                   NaN        
3                                             Practical        
4                                            Self-study        
...                                                 ...        
1195                                        Group-study        
1196                                          Practical        
1197                                          Practical        
1198                                                NaN        
1199                                        Group-study        

      Number_of_absent_days_in_the_current_semester  \
0                                               6.0   
1                                               NaN   
2                                              10.0   
3                                               4.0   
4                

**Виведення перших `10` рядків даних**

Для цього буде використано метод `head` з аргументом `10`:

In [106]:
df.head(10)

Student_identifier_code_for_participation_in_the_survey  \
0                                            ID_2126        
1                                            ID_1860        
2                                            ID_2130        
3                                            ID_2095        
4                                            ID_2044        
5                                            ID_1121        
6                                            ID_1466        
7                                            ID_1330        
8                                            ID_1087        
9                                            ID_2123        

   Age_of_student_at_the_time_of_the_data_collection_process  \
0                                               24.0           
1                                               20.0           
2                                               19.0           
3                                               20.0           
4                                               19.0           
5                                               24.0           
6                                               18.0           
7                                               23.0           
8                                                NaN           
9                                               20.0           

  Grade_level_of_the_participant_in_the_educational_program  \
0                                           Freshman          
1                                             Senior          
2                                             Junior          
3                                          Sophomore          
4                                             Senior          
5                                           Freshman          
6                                             Junior          
7                                          Sophomore          
8                                          Sophomore          
9                                             Senior          

   Hours_spent_studying_on_a_weekly_basis  \
0                                    13.0   
1                                    28.0   
2                                    21.0   
3                                    28.0   
4                                    25.0   
5                                    17.0   
6                                    15.0   
7                                    23.0   
8                                    15.0   
9                                    17.0   

  Preferred_study_method_in_lecture_or_practical_sessions  \
0                                            Lecture        
1                                            Lecture        
2                                                NaN        
3                                          Practical        
4                                         Self-study        
5                                                NaN        
6                                        Group-study        
7                                            Lecture        
8                                            Lecture        
9                                         Self-study        

   Number_of_absent_days_in_the_current_semester  \
0                                            6.0   
1                                            NaN   
2                                           10.0   
3                                            4.0   
4                                            5.0   
5                                            9.0   
6                                            3.0   
7                                            NaN   
8                                           11.0   
9                                            1.0   

   Average_test_score_in_mathematical_subjects  \
0                                        73.44   
1                                        85.28   
2                                       

**Виведення інформації про датасет**

Для цього буде використано метод `info`:

In [107]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 11 columns):
 #   Column                                                      Non-Null Count  Dtype  
---  ------                                                      --------------  -----  
 0   Student_identifier_code_for_participation_in_the_survey     1085 non-null   object 
 1   Age_of_student_at_the_time_of_the_data_collection_process   1079 non-null   float64
 2   Grade_level_of_the_participant_in_the_educational_program   1085 non-null   object 
 3   Hours_spent_studying_on_a_weekly_basis                      1108 non-null   float64
 4   Preferred_study_method_in_lecture_or_practical_sessions     1073 non-null   object 
 5   Number_of_absent_days_in_the_current_semester               1078 non-null   float64
 6   Average_test_score_in_mathematical_subjects                 1092 non-null   float64
 7   Student_participation_in_extracurricular_activities         1080 non-null   object 
 8 

Висновки щодо датасету:
* має `1200` рядків (входжень даних)
* має `5` кількісних показників (індекси колонок: `[1, 3, 5, 6, 9]`). За них відповідає тип даних `float64`.
* 6 текстових показників (всі інші колонки). Тип даних - `object`.
* пропуски є в кожній з колонок. В середньому по `110-120` пропусків (розмір сету *мінус* `Non-Null Count`)

**Виведення статистичних показників датасету**

Для цього буде використано метод `describe`:

In [108]:
df.describe(include=['float64']) # для кількісних даних

Age_of_student_at_the_time_of_the_data_collection_process  \
count                                        1079.000000           
mean                                           20.974977           
std                                             1.976280           
min                                            18.000000           
25%                                            19.000000           
50%                                            21.000000           
75%                                            23.000000           
max                                            24.000000           

       Hours_spent_studying_on_a_weekly_basis  \
count                             1108.000000   
mean                                22.204874   
std                                 10.062070   
min                                  5.000000   
25%                                 14.000000   
50%                                 22.000000   
75%                                 31.000000   
max                                 39.000000   

       Number_of_absent_days_in_the_current_semester  \
count                                    1078.000000   
mean                                        7.088126   
std                                         4.211298   
min                                         0.000000   
25%                                         3.000000   
50%                                         7.000000   
75%                                        11.000000   
max                                        14.000000   

       Average_test_score_in_mathematical_subjects  \
count                                  1092.000000   
mean                                     74.466667   
std                                      14.353203   
min                                      50.100000   
25%                                      61.977500   
50%                                      74.170000   
75%                                      86.902500   
max                                      99.950000   

       Satisfaction_level_with_academic_facilities_in_percentage  
count                                        1080.000000          
mean                                           75.559500          
std                                            14.602437          
min                                            50.070000          
25%                                            63.180000          
50%                                            75.990000          
75%                                            88.505000          
max                                            99.990000

**Висновки** зробимо на прикладі двох колонок: віку студента та годин, скільки він щотижня витрачає на навчання.

**Щодо віку:**
* середній вік (рядок `mean`) дорівнює приблизно `21` року.
* мінімальний вік (рядок `min`) дорівнює `18` рокам.
* максимальний вік (рядок `max`) - `24` рокам.
* `25%` мають вік у `19` років і нижче.
* `50%` - у `21` рік і нижче.
* `75%` - у `23` роки і нижче.

**Щодо щотижневих годин на навчання:**
* в середньому студенти витрачають `22.2` години на тиждень.
* мінімум витрачають `5` годин.
* максимум - `39`.
* `25%` витрачають 14 годин і нижче.
* `50%` - 22 години і нижче.
* `75%` - 31 годину і нижче.

In [109]:
df.describe(include=[object]) # для текстових даних

Student_identifier_code_for_participation_in_the_survey  \
count                                                1085        
unique                                                712        
top                                               ID_1633        
freq                                                    5        

       Grade_level_of_the_participant_in_the_educational_program  \
count                                                1085          
unique                                                  4          
top                                                Senior          
freq                                                  294          

       Preferred_study_method_in_lecture_or_practical_sessions  \
count                                                1073        
unique                                                  4        
top                                            Self-study        
freq                                                  283        

       Student_participation_in_extracurricular_activities  \
count                                                1080    
unique                                                  2    
top                                                    No    
freq                                                  564    

       Preferred_programming_language_for_project_work  \
count                                             1085   
unique                                               5   
top                                             Python   
freq                                               250   

       Plans_for_further_education_or_employment_after_graduation  
count                                                1080          
unique                                                  4          
top                                             Undecided          
freq                                                  287

**Висновки** також зробимо на прикладі двох колонок: ідентифікатора студента та їх професійного рівня.

**Щодо ідентифікатора:**
* непорожніх значень (рядок `count`) тут `1085`. Це значить, що лише `1085` з `1200` студентів мають свої id.
* унікальних значень (`unique`) тут `712`. Виходить, майже кожен код повторюється принаймні двічі.
* найчастіше (`top`) зустрічається ідентифікатор `'ID_1633'`.
* цей ідентифікатор зустрічається аж `5` разів (`freq`).

**Щодо професійного рівня:**
* непорожніх значень тут також `1085`. Це значить, що лише `1085` з `1200` студентів мають дані про свій професійного рівень.
* унікальних значень тут лише `4`. Тобто, маємо лише 4 категорії.
* найчастіше зустрічається рівень `'Senior'`.
* цей ідентифікатор зустрічається аж `294` рази.

# 2. Перейменування стовпців

Задля зручності, стовпці варто перейменувати. Зробити це можна за допомогою методу `rename`, надавши йому словник з новими назвами.

Спершу переглянемо назви наших колонок через атрибут `columns`:

In [110]:
df.columns

Index(['Student_identifier_code_for_participation_in_the_survey',
       'Age_of_student_at_the_time_of_the_data_collection_process',
       'Grade_level_of_the_participant_in_the_educational_program',
       'Hours_spent_studying_on_a_weekly_basis',
       'Preferred_study_method_in_lecture_or_practical_sessions',
       'Number_of_absent_days_in_the_current_semester',
       'Average_test_score_in_mathematical_subjects',
       'Student_participation_in_extracurricular_activities',
       'Preferred_programming_language_for_project_work',
       'Satisfaction_level_with_academic_facilities_in_percentage',
       'Plans_for_further_education_or_employment_after_graduation'],
      dtype='object')

А потім методом `rename` їх перейменуємо:

In [111]:
df.rename(columns = {
    'Student_identifier_code_for_participation_in_the_survey' : 'student_id',
    'Age_of_student_at_the_time_of_the_data_collection_process' : 'student_age',
    'Grade_level_of_the_participant_in_the_educational_program' : 'grade_level',
    'Hours_spent_studying_on_a_weekly_basis' : 'weekly_hours',
    'Preferred_study_method_in_lecture_or_practical_sessions' : 'pref_method',
    'Number_of_absent_days_in_the_current_semester' : 'absent_days',
    'Average_test_score_in_mathematical_subjects' : 'avg_score_math',
    'Student_participation_in_extracurricular_activities' : 'part_in_extra_act',
    'Preferred_programming_language_for_project_work' : 'proj_lang',
    'Satisfaction_level_with_academic_facilities_in_percentage' : 'satisfaction_percentage',
    'Plans_for_further_education_or_employment_after_graduation' : 'after_grad_plans'
}, inplace=True)

Додатковий аргумент `inplace=True` означає, що метод змінить оригінальний фрейм.

Тепер виведемо перші `5` рядків, аби впевнитись, що стовпці було перейменовано:

In [112]:
df.head(5)

student_id  student_age grade_level  weekly_hours pref_method  absent_days  \
0    ID_2126         24.0    Freshman          13.0     Lecture          6.0   
1    ID_1860         20.0      Senior          28.0     Lecture          NaN   
2    ID_2130         19.0      Junior          21.0         NaN         10.0   
3    ID_2095         20.0   Sophomore          28.0   Practical          4.0   
4    ID_2044         19.0      Senior          25.0  Self-study          5.0   

   avg_score_math part_in_extra_act   proj_lang  satisfaction_percentage  \
0           73.44                No        Java                    84.40   
1           85.28               Yes         NaN                    93.20   
2           96.11               Yes  JavaScript                    64.94   
3           57.10                No         C++                    59.65   
4           65.21                No        Java                    55.81   

  after_grad_plans  
0       Employment  
1        Undecided  
2       Employment  
3        Undecided  
4       Employment

Зміни успішні!

# 3. Обробка пропусків у даних

Як ми знаємо, сет містить пропущені дані.

Спершу виведемо кількість пропусків в кожному зі стопців.
Для цього буде використано метод `isnull`, який перевіряє, чи поле має значення `NaN`, а потім метод `sum`, який порахує їхню кількість:

In [113]:
num_missing_values = df.isnull().sum()
num_missing_values

student_id                 115
student_age                121
grade_level                115
weekly_hours                92
pref_method                127
absent_days                122
avg_score_math             108
part_in_extra_act          120
proj_lang                  115
satisfaction_percentage    120
after_grad_plans           120
dtype: int64

Пропущені значення в кількісних стовпцях можна замінити на середні значення по кожному зі стовпців.

Пропуски можна заповнити методом `fillna`.

Середнє значення обраховується методом `mean`. Встановимо аргумент `numeric_only` на `True`, аби рахувало тільки чисельні значення. І зміни буде збережено в оригінальному фреймі.

In [114]:
df.fillna(df.mean(numeric_only=True), inplace=True)
df

student_id  student_age grade_level  weekly_hours  pref_method  \
0       ID_2126         24.0    Freshman          13.0      Lecture   
1       ID_1860         20.0      Senior          28.0      Lecture   
2       ID_2130         19.0      Junior          21.0          NaN   
3       ID_2095         20.0   Sophomore          28.0    Practical   
4       ID_2044         19.0      Senior          25.0   Self-study   
...         ...          ...         ...           ...          ...   
1195    ID_1826         22.0   Sophomore          36.0  Group-study   
1196    ID_1726         24.0   Sophomore          19.0    Practical   
1197        NaN         23.0      Junior          39.0    Practical   
1198    ID_1989         21.0   Sophomore           7.0          NaN   
1199    ID_1371         19.0   Sophomore          20.0  Group-study   

      absent_days  avg_score_math part_in_extra_act   proj_lang  \
0        6.000000       73.440000                No        Java   
1        7.088126       85.280000               Yes         NaN   
2       10.000000       96.110000               Yes  JavaScript   
3        4.000000       57.100000                No         C++   
4        5.000000       65.210000                No        Java   
...           ...             ...               ...         ...   
1195     3.000000       74.466667                No  JavaScript   
1196    14.000000       87.320000               Yes      Python   
1197    13.000000       57.280000                No           R   
1198     2.000000       93.900000               Yes        Java   
1199     3.000000       74.466667               Yes           R   

      satisfaction_percentage  after_grad_plans  
0                     84.4000        Employment  
1                     93.2000         Undecided  
2                     64.9400        Employment  
3                     59.6500         Undecided  
4                     55.8100        Employment  
...                       ...               ...  
1195                  68.4800  Higher education  
1196                  75.5595         Undecided  
1197                  71.1800       Freelancing  
1198                  78.5800       Freelancing  
1199                  95.1400         Undecided  

[1200 rows x 11 columns]

Тепер видалимо усі рядки, які містять хоча б один пропусе за допомогою методу `dropna`. Оригінальний сет змінюємо.

In [115]:
df.dropna(inplace=True)
df

student_id  student_age grade_level  weekly_hours  pref_method  \
0       ID_2126    24.000000    Freshman     13.000000      Lecture   
3       ID_2095    20.000000   Sophomore     28.000000    Practical   
4       ID_2044    19.000000      Senior     25.000000   Self-study   
6       ID_1466    18.000000      Junior     15.000000  Group-study   
8       ID_1087    20.974977   Sophomore     15.000000      Lecture   
...         ...          ...         ...           ...          ...   
1191    ID_1628    22.000000    Freshman     35.000000      Lecture   
1192    ID_1136    21.000000    Freshman     22.204874  Group-study   
1195    ID_1826    22.000000   Sophomore     36.000000  Group-study   
1196    ID_1726    24.000000   Sophomore     19.000000    Practical   
1199    ID_1371    19.000000   Sophomore     20.000000  Group-study   

      absent_days  avg_score_math part_in_extra_act   proj_lang  \
0        6.000000       73.440000                No        Java   
3        4.000000       57.100000                No         C++   
4        5.000000       65.210000                No        Java   
6        3.000000       80.600000                No      Python   
8       11.000000       66.310000                No  JavaScript   
...           ...             ...               ...         ...   
1191     7.088126       81.610000                No      Python   
1192     2.000000       71.760000               Yes         C++   
1195     3.000000       74.466667                No  JavaScript   
1196    14.000000       87.320000               Yes      Python   
1199     3.000000       74.466667               Yes           R   

      satisfaction_percentage  after_grad_plans  
0                     84.4000        Employment  
3                     59.6500         Undecided  
4                     55.8100        Employment  
6                     72.1900  Higher education  
8                     82.6600        Employment  
...                       ...               ...  
1191                  75.5595  Higher education  
1192                  53.9800        Employment  
1195                  68.4800  Higher education  
1196                  75.5595         Undecided  
1199                  95.1400         Undecided  

[643 rows x 11 columns]

Перевіримо кількість пропущених значень тепер:

In [122]:
df.isnull().sum()

student_id                 0
student_age                0
grade_level                0
weekly_hours               0
pref_method                0
absent_days                0
avg_score_math             0
part_in_extra_act          0
proj_lang                  0
satisfaction_percentage    0
after_grad_plans           0
dtype: int64

По нулях!

# 4. Фільтрація даних

Відфільтруємо студентів, чий вік більший за `20` років. Можна використати індексуваня та логічні вирази:

In [116]:
df[df['student_age'] > 20]

student_id  student_age grade_level  weekly_hours  pref_method  \
0       ID_2126    24.000000    Freshman     13.000000      Lecture   
8       ID_1087    20.974977   Sophomore     15.000000      Lecture   
11      ID_1130    21.000000      Junior     23.000000   Self-study   
12      ID_1769    22.000000      Junior     16.000000    Practical   
19      ID_1459    20.974977   Sophomore      9.000000   Self-study   
...         ...          ...         ...           ...          ...   
1187    ID_2109    21.000000      Senior     15.000000    Practical   
1191    ID_1628    22.000000    Freshman     35.000000      Lecture   
1192    ID_1136    21.000000    Freshman     22.204874  Group-study   
1195    ID_1826    22.000000   Sophomore     36.000000  Group-study   
1196    ID_1726    24.000000   Sophomore     19.000000    Practical   

      absent_days  avg_score_math part_in_extra_act   proj_lang  \
0        6.000000       73.440000                No        Java   
8       11.000000       66.310000                No  JavaScript   
11       9.000000       93.120000               Yes  JavaScript   
12       8.000000       61.650000               Yes           R   
19       7.088126       99.770000               Yes         C++   
...           ...             ...               ...         ...   
1187     4.000000       67.170000                No  JavaScript   
1191     7.088126       81.610000                No      Python   
1192     2.000000       71.760000               Yes         C++   
1195     3.000000       74.466667                No  JavaScript   
1196    14.000000       87.320000               Yes      Python   

      satisfaction_percentage  after_grad_plans  
0                     84.4000        Employment  
8                     82.6600        Employment  
11                    76.3600  Higher education  
12                    86.7100         Undecided  
19                    90.8800         Undecided  
...                       ...               ...  
1187                  87.6400       Freelancing  
1191                  75.5595  Higher education  
1192                  53.9800        Employment  
1195                  68.4800  Higher education  
1196                  75.5595         Undecided  

[403 rows x 11 columns]

Тепер студентів, які відвідують більше `20` годин занять на тиждень та ще й мають рівень задоволення більше `80` відсотків.

Для логічного ТА використовується оператор `&`:

In [117]:
df[(df['weekly_hours'] > 20) & (df['satisfaction_percentage'] > 80)]

student_id  student_age grade_level  weekly_hours  pref_method  \
22      ID_1747         24.0   Sophomore     22.204874   Self-study   
28      ID_1975         22.0      Junior     22.000000   Self-study   
32      ID_1957         19.0    Freshman     37.000000      Lecture   
36      ID_1646         24.0   Sophomore     24.000000    Practical   
61      ID_1729         18.0      Senior     21.000000   Self-study   
...         ...          ...         ...           ...          ...   
1133    ID_1806         24.0   Sophomore     25.000000  Group-study   
1149    ID_1072         18.0   Sophomore     37.000000   Self-study   
1170    ID_1011         23.0   Sophomore     31.000000    Practical   
1173    ID_1660         23.0    Freshman     35.000000    Practical   
1178    ID_1465         20.0    Freshman     31.000000  Group-study   

      absent_days  avg_score_math part_in_extra_act   proj_lang  \
22            8.0       52.810000               Yes      Python   
28            4.0       51.660000                No      Python   
32            9.0       69.840000               Yes  JavaScript   
36            7.0       76.500000               Yes           R   
61           11.0       91.110000                No  JavaScript   
...           ...             ...               ...         ...   
1133         13.0       76.090000                No  JavaScript   
1149          7.0       74.466667               Yes         C++   
1170          6.0       85.750000                No  JavaScript   
1173         10.0       74.466667                No        Java   
1178          8.0       72.480000               Yes      Python   

      satisfaction_percentage  after_grad_plans  
22                      93.04         Undecided  
28                      87.68         Undecided  
32                      88.03        Employment  
36                      94.63        Employment  
61                      92.69        Employment  
...                       ...               ...  
1133                    92.29       Freelancing  
1149                    89.60  Higher education  
1170                    90.23        Employment  
1173                    89.88       Freelancing  
1178                    80.07        Employment  

[147 rows x 11 columns]

Тепер буде створено новий датафрейм, який міститиме лише студентів, які обрали мовою для проєкту `Python`, `C++` або `Java`.

In [118]:
languages = ('Python', 'C++', 'Java')   # кортеж назв мов
mask = df['proj_lang'].isin(languages)  # використано метод isin для булеанового маскування оригінального фрейму
py_cpp_java_students = df[mask]         # новий фрейм з відсортованими студентами
py_cpp_java_students

student_id  student_age grade_level  weekly_hours  pref_method  \
0       ID_2126    24.000000    Freshman     13.000000      Lecture   
3       ID_2095    20.000000   Sophomore     28.000000    Practical   
4       ID_2044    19.000000      Senior     25.000000   Self-study   
6       ID_1466    18.000000      Junior     15.000000  Group-study   
19      ID_1459    20.974977   Sophomore      9.000000   Self-study   
...         ...          ...         ...           ...          ...   
1181    ID_1046    18.000000   Sophomore     19.000000  Group-study   
1184    ID_1710    19.000000      Senior     39.000000   Self-study   
1191    ID_1628    22.000000    Freshman     35.000000      Lecture   
1192    ID_1136    21.000000    Freshman     22.204874  Group-study   
1196    ID_1726    24.000000   Sophomore     19.000000    Practical   

      absent_days  avg_score_math part_in_extra_act proj_lang  \
0        6.000000           73.44                No      Java   
3        4.000000           57.10                No       C++   
4        5.000000           65.21                No      Java   
6        3.000000           80.60                No    Python   
19       7.088126           99.77               Yes       C++   
...           ...             ...               ...       ...   
1181     9.000000           81.10                No    Python   
1184     1.000000           59.62               Yes      Java   
1191     7.088126           81.61                No    Python   
1192     2.000000           71.76               Yes       C++   
1196    14.000000           87.32               Yes    Python   

      satisfaction_percentage  after_grad_plans  
0                     84.4000        Employment  
3                     59.6500         Undecided  
4                     55.8100        Employment  
6                     72.1900  Higher education  
19                    90.8800         Undecided  
...                       ...               ...  
1181                  96.3800  Higher education  
1184                  70.1400       Freelancing  
1191                  75.5595  Higher education  
1192                  53.9800        Employment  
1196                  75.5595         Undecided  

[403 rows x 11 columns]

# 5. Групування даних

Групування даних у Pandas дуже схоже до SQL. Виконати його можна за допомогою методу `groupby`.

Спочатку згрупуємо студентів за їхнім професійним рівнем:

In [119]:
grouped_data = df.groupby('grade_level')
grouped_data

Тепер методом `mean` можна порахувати середню кількість годин, які кожна група витрачала на навчання, попередньо обравши колонку `'weekly_hours'`:

In [120]:
grouped_data['weekly_hours'].mean()

grade_level
Freshman     22.615853
Junior       22.260057
Senior       22.845374
Sophomore    22.557674
Name: weekly_hours, dtype: float64

За допомогою методу `apply` та наданій йому лямбда-функції, яка обраховує кількість `'Yes'` у колонці `'part_in_extra_act'`, можна порахувати кількість студентів у кожній групі, які беруть участь у позанавчальних активностях:

In [121]:
grouped_data['part_in_extra_act'].apply(lambda x: (x == 'Yes').sum())

grade_level
Freshman     74
Junior       62
Senior       88
Sophomore    81
Name: part_in_extra_act, dtype: int64

# Висновок

У результаті виконання даної лабораторної роботи я:
* навчився працювати з бібліотекою Pandas
* навчився створювати датафрейми
* аналізувати надану в них інформацію
* працювати з пропущеними даними
* фільтрувати їх, а також групувати.

Мною було написано працюючий код, усі результати виконання якого (разом з описами) представлено вище.